src/components/text-editor/prosemirror-adapter/plugins/trigger/inserter.ts

```diff
@@ -1,44 +1,80 @@
-import { Node, NodeType, Schema } from 'prosemirror-model';
+import { Node, NodeType, Schema, DOMParser, Fragment } from 'prosemirror-model';
 import { EditorView } from 'prosemirror-view';
 import {
     TextEditor,
     TextEditorNode,
 } from 'src/components/text-editor/text-editor.types';
+import { ContentTypeConverter } from '../../../utils/content-type-converter';
 
-export const inserterFactory = (view: EditorView): TextEditor => {
+export const inserterFactory = (
+    view: EditorView,
+    contentConverter: ContentTypeConverter,
+): TextEditor => {
     const startPos = getTriggerStartPosition(view);
 
     return {
-        insert: (input: TextEditorNode | string) => {
-            const { state, dispatch } = view;
-            const { schema, selection } = state;
-            const { $from } = selection;
+        insert: createNodeAndTextInserter(view, startPos),
+        insertHtml: createHtmlInserter(view, contentConverter, startPos),
+    };
+};
+
+const createNodeAndTextInserter =
+    (view: EditorView, startPos: number) =>
+    (input: TextEditorNode | string): void => {
+        const schema = view.state.schema;
+        let node: Node;
 
-            let node: Node;
+        try {
+            node = createNode(input, schema);
+        } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(error.message);
 
-            try {
-                node = createNode(input, schema);
-            } catch (error) {
-                // eslint-disable-next-line no-console
-                console.error(error.message);
+            return;
+        }
 
-                return;
-            }
+        const spaceNode = schema.text(' ');
 
-            const spaceNode = schema.text(' ');
+        const fragment = schema.nodes.doc.create(null, [node, spaceNode]);
 
-            const fragment = schema.nodes.doc.create(null, [node, spaceNode]);
+        dispatchTransaction(view, startPos, fragment);
+    };
 
-            const transaction = state.tr.replaceWith(
-                startPos,
-                $from.pos,
-                fragment,
+const createHtmlInserter =
+    (
+        view: EditorView,
+        contentConverter: ContentTypeConverter,
+        startPos: number,
+    ) =>
+    (input: string): void => {
+        const schema = view.state.schema;
+
+        (async () => {
+            const container = document.createElement('span');
+            container.innerHTML = await contentConverter.parseAsHTML(
+                input,
+                schema,
             );
 
-            transaction.setMeta('stopTrigger', true);
-            dispatch(transaction);
-        },
+            const fragment =
+                DOMParser.fromSchema(schema).parse(container).content;
+
+            dispatchTransaction(view, startPos, fragment);
+        })();
```

Comment:
> But, I don't understand what the difference between `insert` and `insertHtml` is supposed to be? Both can insert strings, so what's the difference?

OK, I'm reading the code now, so I might find an answer to this myself anyway. I'll return with the same question if I'm not convinced when I'm done reading the code 😄

Reply:
Nope, I still don't understand why we need to separate functions. Not sure if I'm missing something… 🙂

- If we can't trust the consumer with strings injected via `insertHtml`, I don't see why we would trust them with strings injected via `insert`.
- Conversely, we want to render any html that is injected via `insertHtml`, but if someone inserts html via `insert`, do we not want to render that as html?

Reply:
The difference between `insert` and `insertHtml` is that `insert` doesn't parse the string as html, which we want for the email story.

`insertHtml` will parse and sanitize the input string.

![Screenshot 2024-11-27 at 12 19 33](https://github.com/user-attachments/assets/e647c87e-7a3e-4770-a330-1571acf72b39)

Reply:
Yes exactly!

Reply:
Then my question (to @civing and @john-traas I guess) is:

Don't we want that for strings inserted via `insert` too?

Reply:
Since we wanted to sanitise the html, I assume we don't trust the consumer (or, at least, we don't trust that the consumer will _definitely_ sanitise the input for us). Why then would we trust that the consumer doesn't send us something nasty via the. `insert` function, just because we asked nicely? 😄

---

Ah, this PR writes the input to an element and then uses Prosemirror's `DOMParser` to read it into Prosemirror's state… I didn't notice that before…

OK, I can understand that you'd want to avoid that. But if that's the only way to do it, I think it's still worth it to simplify the interface, yes. Assuming we only go this route when we insert a mention or a text snippet or something like that, and this isn't being invoked for what's already in Prosemirror every time you type a letter, or something insane like that?

Reply:
I'm guessing `DOMParse` can handle our custom components? If so, one benefit of this unified route is that we can kill the `createNode` function 😊

Reply:
Can you try if it works for all the cases @FredrikWallstrom ? To just use your function for everything?

Reply:
That would be neat, I will try!

Reply:
But it works in the initial parsing right? 🤔 
Meaning we can have a custom element like <limebb-mention /> and still have it rendered only relying on that initial parsing which are the DOM Parser from prosemirror.

Reply:
I don't think so, because we need to create and insert that specialised mention node so that the editor view recognises that `this is a mention node` and then either serialises or parses it using our serialiser/parser.

Reply:
Thank you John. I will stay low here, please let me know if I should do something.

Reply:
> I'm guessing `DOMParse` can handle our custom components? If so, one benefit of this unified route is that we can kill the `createNode` function 😊

The reason the DOM parser can handle our custom components is because of the custom node that tells it how handle it. We create and insert the relevant node with that information. If we remove the `createNode` function then it won't work.